### PR TITLE
New language implementation reference

### PR DESCRIPTION
**Pull request by dburriss**
This PR addresses part of a discussion in issue #359 to provide developer guidance on how to support new languages and implement metrics for those languages.

Was initially part of PR #686 but that PR is split into multiple and closed.

This PR includes the docs for setting up of a new grammar and a start at metrics. Metrics docs may be split out into their own MD file later.

### CODE REVIEW

**Comment by dburriss** [comment]
@Luni-4 would this alleviate some of your maintenance concerns about optimizing for the reader rather than the maintainer? https://github.com/dburriss/rust-code-analysis/blob/dburriss/feature/doc-link-check/link-check.sh
The script will verify all links as well as any file paths that may be found in the markdown link description part.

File: rust-code-analysis-book/src/developers/loc.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    (1..=num).product()
+}
+```
+
+The example above will be used to illustrate each of the **LoC** metrics described below.
+
+### SLOC
+
+A straight count of all lines in the file including code, comments, and blank lines.  
+METRIC VALUE: 5
```
[the remark this answers is not in the document]
Comment (dburriss) [reply rust-code-analysis-book/src/developers/loc.md:26]: 🤦 I went and added comments after and forgot to update the value.

**Comment by Luni-4** [comment]
> @Luni-4 would this alleviate some of your maintenance concerns about optimizing for the reader rather than the maintainer? https://github.com/dburriss/rust-code-analysis/blob/dburriss/feature/doc-link-check/link-check.sh The script will verify all links as well as any file paths that may be found in the markdown link description part.

That's a good script, but I don't think it is necessary, I mean, I'm not against the links in the documentation, that's not the problem, but I'm against that kind of information that needs to be checked and updated constantly (like the version of tree-sitter). In the review comments above I was able to solve that problem though maintaining your approach too, so let's resolve all the conversation above and I think we are ready to go. We will iterate over that in a following PR or later. What do you think?

Thanks for your effort anyway! :)

**Comment by dburriss** [comment]
Hi @Luni-4 
You don't know me but I am generally regarded as patient and polite, so I really hope that this comment comes off in the spirit it is meant. The tone is often lost in comments.
In my job as CTO and other positions over the last few years, I do not find time to code at work often. So I have my own little projects that I have a few hours a week to code on between spending time with family and friends.
I really! really! do appreciate the time you spent reviewing and commenting on this PR. Wherever there were corrections to be made, I am grateful for you catching them. Where my knowledge was lacking, I am grateful for the knowledge you shared.
There are 2 points we are getting hung up on and I wanted to be explicit about them so we can find a solution.

**Explicit reference to files for changes or examples**  
As someone who is new to the project and wrote this while trying to figure it out, I am convinced this is what is best for a new contributor. I do acknowledge the maintenance concern. To alleviate this concern I suggest we PR [this script file](https://github.com/dburriss/rust-code-analysis/blob/dburriss/feature/doc-link-check/link-check.sh) in that can be used to verify the integrity of the paths and links in the docs.

**Style**  
Many of the requested changes are just a matter of style. For some of these, I acknowledged that the structure was better (and I made the change). Some I was doing to just try to get this PR through. I have stopped doing that as it was using my limited time to add no value.  
For this one, I see a few options.

1. We somehow make the required style of the documentation explicit. Honestly, this seems unnecessarily restrictive but at least the constraints would be clear.
2. Style changes are seen completely as suggestions but we mark them as such.
3. The people who get to dictate the style write the documentation.
4. Documentation content (like this one) can go through correct and style changes can be done later if someone has a strong enough desire to PR a change.

Anyway, I didn't want to just abandon this PR without an explanation. And to be clear, I do not want to abandon it. For my own mental health though I cannot be spending both my work time and my free time only working on things that bring me no joy.  
This started as me wanting to implement metrics for other languages that I need support for on a hobby project and we are entering week 2 on reviewing 2 small markdown files. Unfortunately, I cannot continue not knowing where it will end. I had planned to write up the other metrics docs too but with this workflow, I don't think that is feasible.

**Comment by Luni-4** [comment]
> Hi @Luni-4 You don't know me but I am generally regarded as patient and polite, so I really hope that this comment comes off in the spirit it is meant. The tone is often lost in comments. In my job as CTO and other positions over the last few years, I do not find time to code at work often. So I have my own little projects that I have a few hours a week to code on between spending time with family and friends. I really! really! do appreciate the time you spent reviewing and commenting on this PR. Wherever there were corrections to be made, I am grateful for you catching them. Where my knowledge was lacking, I am grateful for the knowledge you shared. There are 2 points we are getting hung up on and I wanted to be explicit about them so we can find a solution.

Hi @dburriss! 
First of all, sorry if my tone seemed rude in the last comments, I didn't want to show that attitude. So don't take the rest of this comment as accusatory. And you have been polite, don't worry! 
Anyway, I understand your points about the time and even I would have preferred not to do a so in-depth review, but I want to explain why I have done that. 

Maintaining the documentation is hard, sometimes you receive comments like "I don't understand that sentence, could you please change it?" or something like "Why have you used that version? The current code is quite different compared to what it is written the documentation!" and so on. 

So in this review I tried to be more precise as possible to avoid that kind of problems later in the time. That's why I requested you to make all those style changes. A not precise documentation usually leads to a series of subsequently PRs to fix those problems and so a maintenance burden for the collaborators of a project. My attitude was not intended to mine your work, absolutely! 

About your script, I like that, but I think it is not necessary here because your paths are almost permalinks (Cargo.toml, language.rs), so we can use your script, but not for this PR.

> **Explicit reference to files for changes or examples** As someone who is new to the project and wrote this while trying to figure it out, I am convinced this is what is best for a new contributor. I do acknowledge the maintenance concern. To alleviate this concern I suggest we PR [this script file](https://github.com/dburriss/rust-code-analysis/blob/dburriss/feature/doc-link-check/link-check.sh) in that can be used to verify the integrity of the paths and links in the docs.

I agree with you about the links, indeed I haven't removed the paths and links in docs (if I had done that, please point me out to those comments, because I made a mistake then)

> **Style** Many of the requested changes are just a matter of style. For some of these, I acknowledged that the structure was better (and I made the change). Some I was doing to just try to get this PR through. I have stopped doing that as it was using my limited time to add no value. For this one, I see a few options.
> 
>     1. We somehow make the required style of the documentation explicit. Honestly, this seems unnecessarily restrictive but at least the constraints would be clear.
> 
>     2. Style changes are seen completely as suggestions but we mark them as such.
> 
>     3. The people who get to dictate the style write the documentation.
> 
>     4. Documentation content (like this one) can go through correct and style changes can be done later if someone has a strong enough desire to PR a change.

The points above are always related to what I said above, the style changes (so the rephrasing of sentences) are suggested in the spirit of avoid annoying requests later in the time. Obviously we can change that later, but who is going to do that? And when? That's why I think it is better to be reeeeally precise beforehand and not later

> Anyway, I didn't want to just abandon this PR without an explanation. And to be clear, I do not want to abandon it. For my own mental health though I cannot be spending both my work time and my free time only working on things that bring me no joy. This started as me wanting to implement metrics for other languages that I need support for on a hobby project and we are entering week 2 on reviewing 2 small markdown files. Unfortunately, I cannot continue not knowing where it will end. I had planned to write up the other metrics docs too but with this workflow, I don't think that is feasible.

In my opinion we can solve this matter in this way. If you don't want to spend your time in writing the documentation, but code, feel free to use your time to code! So don't feel obligated to write the documentation if you don't want. We are having fun here! :smiley: 
But please, let's just finish this PR (so address the remaining comments above) and then you can use your time to do the Java metrics and all the other metrics. Thanks in advance for your effort in completing this PR!

**Comment by marco-c** [comment]
>  But please, let's just finish this PR (so address the remaining comments above) and then you can use your time to do the Java metrics and all the other metrics. Thanks in advance for your effort in completing this PR!

We could merge the PR as-is if it looks good enough, and file a follow-up issue to fix the remaining comments.
This way @dburriss can move on to something he enjoys more :)

**Review by marco-c — APPROVED** [review]

**Comment by dburriss** [comment]
@marco-c @Luni-4 although I am glad this is merged, I had intended to relent and make all changes. Unfortunately, a tragedy in the family meant this moved very far down in my list of priorities. I will move on to the LoC metric.

**Comment by marco-c** [comment]
@dburriss I'm really sorry for that!
BTW, I've made most of the changes myself manually, it was just quicker to do that and merge the (very useful) PR.

**Pull request merged**